### PR TITLE
Add public view of card grant spending 

### PR DESCRIPTION
**Pull request by sampoder**
This uses a model that pops up on transparent orgs with card grants:

Screenshot 2023-11-01 at 12 04 44 AM

All private info should be hidden, though I'd appreciate a sanity check on that.

I thought a model was the cleanest way to do this.

Closes #4380.

Happy halloween! 🎃

### CODE REVIEW

**Comment by maxwofford** [comment]
Awesome work on this @sampoder! I'm half tempted to merge even with it needing tweaks because this by itself makes everything so much clearer.

**Comment by sampoder** [comment]
@maxwofford awesome! i just changed it up a little so that this view shows up when you click "Details" while signed in as an organiser and added a little icon to keep it visually distinct

File: app/assets/javascripts/ui.js
Comment (YodaLightsabr) [inline app/assets/javascripts/ui.js]: I think this might interfere with or break the HcbCode popover implementation
https://github.com/hackclub/hcb/blob/837ba812492346d041c5eddf50865b9941097c3d/app/models/hcb_code.rb#L53

File: app/controllers/card_grants_controller.rb
Comment (YodaLightsabr) [inline app/controllers/card_grants_controller.rb]:
```suggestion
    @frame = params[:frame].present?
```

Thoughts on using `frame` instead of `async_frame`?

File: app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb
Comment (YodaLightsabr) [inline app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb]:
```suggestion
  <% if local_assigns[:receipt_upload_button] %>
```
This might be a little bit cleaner (for cases like this where you want to be able to pass in `false`)

File: app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb
Comment (YodaLightsabr) [inline app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb]:
```suggestion
  <% if local_assigns[:receipt_upload_button] %>
```

File: app/views/canonical_transactions/_canonical_transaction.html.erb
Comment (YodaLightsabr) [inline app/views/canonical_transactions/_canonical_transaction.html.erb]:
```suggestion
  <% if local_assigns[:receipt_upload_button] %>
```

File: app/views/canonical_transactions/_canonical_transaction.html.erb
Comment (YodaLightsabr) [inline app/views/canonical_transactions/_canonical_transaction.html.erb]:
```suggestion
  <% if local_assigns[:receipt_upload_button] %>
```

**Review by YodaLightsabr — COMMENTED** [review]
Nice work on this! I just wanted to add a few comments to make sure our popover implementation is consistent.

File: app/assets/javascripts/ui.js
Comment (sampoder) [reply app/assets/javascripts/ui.js]: that explains a lot omg... lol yes, let's use frame

Comment (sampoder) [reply app/assets/javascripts/ui.js]:
```suggestion
      $.get(`${$(frame).data('src')}?frame=true`,  data => {
```

File: app/assets/javascripts/ui.js
Comment (sampoder) [inline app/assets/javascripts/ui.js]:
```suggestion
      $.get($(frame).data('src')`,  data => {
```

File: app/assets/javascripts/ui.js
Comment (sampoder) [inline app/assets/javascripts/ui.js]:
```suggestion
      $.get($(frame).data('src'),  data => {
```

File: app/views/canonical_transactions/_canonical_transaction.html.erb
Comment (sampoder) [inline app/views/canonical_transactions/_canonical_transaction.html.erb]:
```suggestion
    <%= async_frame_to public_card_grant_path(ct.local_hcb_code.disbursement.card_grant, params: { frame: true }), as: :div do %>
```

**Comment by sampoder** [comment]
@maxwofford @YodaLightsabr: whilst the HCPD may be fictitious, what do we need to do to push this PR over the line? Does it look good to the both of you?

File: app/controllers/card_grants_controller.rb
Comment (garyhtou) [inline app/controllers/card_grants_controller.rb]: This action should be renamed. `public` is basically like a keyword in Ruby; we don't want to be overriding that

File: app/views/canonical_transactions/_canonical_transaction.html.erb
Comment (garyhtou) [inline app/views/canonical_transactions/_canonical_transaction.html.erb]: Do we wanna update the transfer page to have the same terminology/icon?

https://github.com/hackclub/bank/blob/c1807ec9e56164a6fca4a1c9b4dfe35aea440340/app/views/events/transfers.html.erb#L81-L83

File: config/routes.rb
Comment (garyhtou) [inline config/routes.rb]: if we have `get "public"` at the bottom, I don't think we need `:public` here

File: app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb
Comment (garyhtou) [inline app/views/canonical_pending_transactions/_canonical_pending_transaction.html.erb]: There is no need to check if `@force_no_popover` is defined. Undefined instance variables return `nil`. So, `(!defined?(@force_no_popover) || !@force_no_popover)` I think can be rewritten as just `!@force_no_popover`

File: app/views/canonical_transactions/_canonical_transaction.html.erb
Comment (garyhtou) [inline app/views/canonical_transactions/_canonical_transaction.html.erb]: Is the only difference the link? If so, do

```ruby
<% details_link = ct.local_hcb_code.card_grant? ? public_card_grant_path(ct.local_hcb_code.disbursement.card_grant) : ct.local_hcb_code.url %>

<%= link_to "Details", details_link, blath blath blah other stuff %>
```

File: app/views/card_grants/show.html.erb
Comment (garyhtou) [inline app/views/card_grants/show.html.erb:1]: Instead of using the `show` view for both the full-page and "public"/popover version. Let's create a separate page for the public version. It can sit at `card_grants/:id/popover`.

Over all shared content from this page into their own partials. I'm thinking we could have 3:
- `balance`
- `spending_instructions`
- `transactions`

This will also us to get rid of the `@public` instance variable

**Review by garyhtou — CHANGES_REQUESTED** [review]
This is very cool! Two large feedbacks:
- We should find a new name for the controller action (we're currently overriding an existing`public` method)
- Give the popover (public) version its own view file. This means moving shared content into partials.

**Review by maxwofford — DISMISSED** [review]
Hey @sampoder! You or @garyhtou should have a chat with @SerenityUX. HCPD might not be real, but the HCPD: Pizza Crimes Division is a real thing headed by them and they probably have some opinions about this!

As for https://github.com/hackclub/hcb/issues/4380, I think this is missing one big component: on a regular event, I can audit it by looking at the transaction history in linear order. If I check it each day, I can look at the most recent ~5 transactions and stop searching once I reach the previous day's transactions.

On card grant events, if I want to see if someone has done something invalid in the past day, **I have to go through every card's popover every single day**. I think that's too hard to audit / call transparent.

**Comment by sampoder** [comment]
howdy! that makes sense to me. i guess that wasn't quite the problem i was trying to solve with the original PR which is why what we've got here isn't a perfect fit. instead of trying to shoehorn in a fix to what we already have, would it make sense to build out an admin tool like we have with https://github.com/hackclub/hcb/pull/4355? some other ideas I have are: including card grant transactions in exports or having a toggle to list them on the ledger directly.

**Comment by sampoder** [comment]
Oh, and I'll also text Thomas!

**Comment by sarthaktexas** [comment]
What you suggested (daily compliance checks) is what Thomas, Jasper, and I agreed on. I think this PR is fine to merge especially because they need this quickly. cc @maxwofford

**Comment by sarthaktexas** [comment]
@maxwofford said his request to change was in fact not a request to not merge. We are just waiting for an approval from @garyhtou.

**Comment by garyhtou** [comment]
> What you suggested (daily compliance checks) is what Thomas, Jasper, and I agreed on.

Is this their intention behind https://github.com/hackclub/hcb/issues/4932?

**Comment by sarthaktexas** [comment]
Oh we have two options. Webhook works if you want to add that into the codebase -- or what Max suggested is a compliance check feature for organizations (not an internal admin tool) giving out card grants.

**Comment by garyhtou** [comment]
> Oh we have two options. Webhook works if you want to add that into the codebase -- or what Max suggested is a compliance check feature for organizations (not an internal admin tool) giving out card grants.

I think webhooks will be a lighter lift for us. Let's go for that especially since it's a very niche use case (we have few card grant programs). https://github.com/hackclub/hcb/issues/4932 outlines the spec that Jasper and I agreed on

**Review by garyhtou — APPROVED** [review]
coolio!

**Pull request merged**
